upstream/man/tcamgr.1
.TH "TCATEST" 3 "2010-08-05" "Man Page" "Tokyo Cabinet"

.SH NAME
tcamgr \- the command line utility of the abstract database API

.SH DESCRIPTION
.PP
The command `\fBtcamgr\fR' is a utility for test and debugging of the abstract database API and its applications.  `\fIname\fR' specifies the name of a database.  `\fIkey\fR' specifies the key of a record.  `\fIvalue\fR' specifies the value of a record.  `\fIparams\fR' specifies the tuning parameters.  `\fIfunc\fR' specifies the name of a function.  `\fIarg\fR' specifies the arguments of the function.  `\fIdest\fR' specifies the path of the destination file.
.PP
.RS
.br
\fBtcamgr create \fIname\fB\fR
.RS
Create a database file.
.RE
.br
\fBtcamgr inform \fIname\fB\fR
.RS
Print miscellaneous information to the standard output.
.RE
.br
\fBtcamgr put \fR[\fB\-sx\fR]\fB \fR[\fB\-sep \fIchr\fB\fR]\fB \fR[\fB\-dk\fR|\fB\-dc\fR|\fB\-dai\fR|\fB\-dad\fR]\fB \fIname\fB \fIkey\fB \fIvalue\fB\fR
.RS
Store a record.
.RE
.br
\fBtcamgr out \fR[\fB\-sx\fR]\fB \fR[\fB\-sep \fIchr\fB\fR]\fB \fIname\fB \fIkey\fB\fR
.RS
Remove a record.
.RE
.br
\fBtcamgr get \fR[\fB\-sx\fR]\fB \fR[\fB\-sep \fIchr\fB\fR]\fB \fR[\fB\-px\fR]\fB \fR[\fB\-pz\fR]\fB \fIname\fB \fIkey\fB\fR
.RS
Print the value of a record.
.RE
.br
\fBtcamgr list \fR[\fB\-sep \fIchr\fB\fR]\fB \fR[\fB\-m \fInum\fB\fR]\fB \fR[\fB\-pv\fR]\fB \fR[\fB\-px\fR]\fB \fR[\fB\-fm \fIstr\fB\fR]\fB \fIname\fB\fR
.RS
Print keys of all records, separated by line feeds.
.RE
.br
\fBtcamgr optimize \fIname\fB \fIparams\fB\fR
.RS
Optimize a database file.
.RE
.br
\fBtcamgr misc \fR[\fB\-sx\fR]\fB \fR[\fB\-sep \fIchr\fB\fR]\fB \fR[\fB\-px\fR]\fB \fIname\fB \fIfunc\fB \fR[\fB\fIarg\fB...\fR]\fB\fR
.RS
Call a versatile function for miscellaneous operations.
.RE
.br
\fBtcamgr map \fR[\fB\-fm \fIstr\fB\fR]\fB \fIname\fB \fIdest\fB\fR
.RS
Map records into another B+ tree database.
.RE
.br
\fBtcamgr version\fR
.RS
Print the version information of Tokyo Cabinet.
.RE
.RE
.PP
Options feature the following.
.PP
.RS
\fB\-sx\fR : the input data is evaluated as a hexadecimal data string.
.br
\fB\-sep \fIchr\fR\fR : specify the separator of the input data.
.br
\fB\-dk\fR : use the function `tcadbputkeep' instead of `tcadbput'.
.br
\fB\-dc\fR : use the function `tcadbputcat' instead of `tcadbput'.
.br
\fB\-dai\fR : use the function `tcadbaddint' instead of `tcadbput'.
.br
\fB\-dad\fR : use the function `tcadbadddouble' instead of `tcadbput'.
.br
\fB\-px\fR : the output data is converted into a hexadecimal data string.
.br
\fB\-pz\fR : do not append line feed at the end of the output.
.br
\fB\-m \fInum\fR\fR : specify the maximum number of the output.
.br
\fB\-pv\fR : print values of records also.
.br
\fB\-fm \fIstr\fR\fR : specify the prefix of keys.
.br
.RE
.PP
This command returns 0 on success, another on failure.

.SH SEE ALSO
.PP
.BR tcatest (1),
.BR tcamttest (1),
.BR tcadb (3),
.BR tokyocabinet (3)


upstream/ChangeLog
2009-12-07  Mikio Hirabayashi  <hirarin@gmail.com>

	* tctdb.c (tctdbidxputtoken, tctdbidxputqgram): domain overflow checking was added.

	- Release: 1.4.41

2009-11-24  Mikio Hirabayashi  <hirarin@gmail.com>

	* tchdb.c (tchdbputimpl): efficiency of using the free block pool was improved.

	* tchdb.c (tchdboptimizeimpl): the behaviour in no lock mode was modified.

	- Release: 1.4.40

2009-11-11  Mikio Hirabayashi  <hirarin@gmail.com>

	* tcutil.h (TCALIGNOF, tcgeneric_t): new macros.

	* tcutil.h (TCALIGNPAD): the alignment is now calculated by the generic union.

	* myconf.h (_alignof, _issigned, _maxof): new macros.

	- Release: 1.4.39

2009-11-09  Mikio Hirabayashi  <hirarin@gmail.com>

	* tcutil.c (tcsysinfo): FreeBSD and Mac OS X are now supported.

	* tcutil.c (tcmpoolclear): new function.

	* tchdb.c (tchdbputproc): a potential bug of memory aligment violation was fixed.

	* tcbdb.c (tcbdbputproc): a potential bug of memory aligment violation was fixed.

	* tcfdb.c (tcfdbputproc): a potential bug of memory aligment violation was fixed.

	- Release: 1.4.38

2009-10-27  Mikio Hirabayashi  <hirarin@gmail.com>

	* tctdb.c (tctdbcacheclear, tctdbcacheclearimpl): new functions.

	* tcadb.c (tcadbmisc): "getpart", "cacheclear", and "regex" functions were added.

	- Release: 1.4.37

2009-10-26  Mikio Hirabayashi  <hirarin@gmail.com>

	* tcutil.c (tcwwwformdecode2): a bug related to multiple cookie handling was fixed.

	* tcadb.c (tcadbmisc): the "error" functions were added.

	- Release: 1.4.36

2009-10-06  Mikio Hirabayashi  <hirarin@gmail.com>

	* tchdb.c (tchdbseekwrite): a bug related to emulation of the UBC was fixed.

	* tchdb.c (tchdbvanishimpl): potential memory leak was removed.

	* tcfdb.c (tcfdbvanishimpl): potential memory leak was removed.

	* tcadb.c (tcadbsetskelmulti, tcadbmulnew, tcadbmuldel): new functions.

	* tcadb.c (tcadbmisc): "range" sub function for B+ tree was added.

	* tcamgr.c (setskeltran): new function.

	* tcamttest.c: new file.

	- Release: 1.4.35

2009-09-11  Mikio Hirabayashi  <hirarin@gmail.com>

	* myconf.h: a macro for environment without "nanl" was added.

	* tctdbmetasearch (tctdbmetasearch): a bug related to the limit parameters was fixed.

	- Release: 1.4.34

2009-08-31  Mikio Hirabayashi  <hirarin@gmail.com>

	* tcutil.h, tchdb.h, tcbdb.h, tcfdb.h, tctdb.h, tcadb.h: including stdbool.h was wrapped.

	* tcutil.c (tccmpdecimal): accuracy was improved and now real number is supported.

	* tctdb.c (tctdbqrycondmatch): accuracy of decimal operators was improved.

	- Release: 1.4.33

2009-08-14  Mikio Hirabayashi  <hirarin@gmail.com>

	* tcutil.c (tctmpldumpeval): "INC", "PRT" operators and "SET" directive were added.

	* tcutil.c (tcwwwformdecode2): a bug of memory corruption was fixed.

	* tchdb.c (tchdbgetimpl): a bug related to caching large records was fixed.

	- Release: 1.4.32

2009-07-23  Mikio Hirabayashi  <hirarin@gmail.com>

	* tcutil.c (tcmpoolpop, tcstatfile): new functions.

	* tcutil.c (tcwwwformdecode2, tcarccipher): new functions.

	* tcutil.c (tctmpldumpeval, tctmpldumpevalvar): type checking was improved.

	- Release: 1.4.31

2009-07-13  Mikio Hirabayashi  <hirarin@gmail.com>

	* tctdb.c (tctdbstrtometasearcytype): new function.

	* tcadb.c (tcadbmisc): "metasearch" command was added.

	- Release: 1.4.30

2009-07-03  Mikio Hirabayashi  <hirarin@gmail.com>

	* tctdb.c (tctdbidxgetbyftsunion): efficiency counting sort was improved.

	- Release: 1.4.29

2009-06-22  Mikio Hirabayashi  <hirarin@gmail.com>

	* tcutil.c (tcstrutfnorm, tcstrkwic, tcstrtokenize): new functions.

	* tcutil.c (tcstrutfnorm, tcstrucsnorm): "TCUNWIDTH" option was added.

	* tctdb.c (tctdbsetinvcache, tctdbidxsyncicc, tctdbidxcmpkey): new functions.

	* tctdb.c (tctdbmetasearch, tctdbget4, tctdbqrykwic): new functions.

	* tctdb.c (tctdbsetindeximpl): the xmsiz parameter is now configured.

	- Release: 1.4.28

2009-06-16  Mikio Hirabayashi  <hirarin@gmail.com>

	* tcutil.c (tcstrskipspc, tcstrucsnorm): new functions.

	* tctdb.c (tctdbqryaddcond): full-text search operators were added.

	* tctdb.c (tctdbsetindex): q-gram inverted index was added.

	* tctdb.c (tctdbidxputqgram, tctdbidxoutqgram, tctdbidxgetbyfts): new functions.

	- Release: 1.4.27

2009-06-13  Mikio Hirabayashi  <hirarin@gmail.com>

	* tctdb.c (tctdbsetindex, tctdbsearchimpl): token inverted index was added.

	* tctdb.c (tctdbidxputone, tctdbidxoutone): new functions.

	* tctdb.c (tctdbidxputtoken, tctdbidxouttoken, tctdbidxgetbytokens): new functions.

	* tcadb.c (tcadbput, tcadbputkeep, tcadbputcat): capnum behaviour was modified.

	- Release: 1.4.26

2009-06-09  Mikio Hirabayashi  <hirarin@gmail.com>

	* tcutil.c (tccstrescape, tccstrunescape, tcjsonescape, tcjsonunescape): new functions.

	* tcutil.c (tcpathlock, tcpathunlock): new functions.

	* tchdb.c (tchdbopen): double opening detection was added.

	* tchdb.c (tchdbtranbegin): a bug of invalid flag setting was fixed.

	* tcfdb.c (tcfdbopen): double opening detection was added.

	* tcfdb.c (tcfdbtranbegin): a bug of invalid flag setting was fixed.

	* tctdb.c (tctdbiternext3): new function.

	- Release: 1.4.25

2009-06-04  Mikio Hirabayashi  <hirarin@gmail.com>

	* tcutil.c (tcwwwformencode, tcwwwformdecode): new functions.

	* tcutil.c (tctmplnew, tctmpldel, tctmplload, tctmpldump): new functions.

	* tcutil.c (tcmapget4, tctreeget4): new functions.

	* tcutil.c (tclistprintf, tcmapprintf, tctreeprintf): new functions.

	* tcucodec.c (runtmpl, proctmpl): new functions.

	- Release: 1.4.24

2009-05-24  Mikio Hirabayashi  <hirarin@gmail.com>

	* tcutil.c (tcmapiterinit2, tcmdbiterinit2): new functions.

	* tchdb.c (tchdbiterinit2, tchdbiterjumpimpl): new functions.

	* tcfdb.c (tcfdbiterinit2, tcfdbiterjumpimpl): new functions.

	* tctdb.c (tctdbiterinit2): new function.

	- Release: 1.4.23

2009-05-21  Mikio Hirabayashi  <hirarin@gmail.com>

	* tcadb.c (tcadbmisc): "sync", "optimize", "vanish" commands were added.

	- Release: 1.4.22


upstream/tcamgr.c
/*************************************************************************************************
 * The command line utility of the abstract database API
 *                                                      Copyright (C) 2006-2008 Mikio Hirabayashi
 * This file is part of Tokyo Cabinet.
 * Tokyo Cabinet is free software; you can redistribute it and/or modify it under the terms of
 * the GNU Lesser General Public License as published by the Free Software Foundation; either
 * version 2.1 of the License or any later version.  Tokyo Cabinet is distributed in the hope
 * that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Lesser General Public
 * License for more details.
 * You should have received a copy of the GNU Lesser General Public License along with Tokyo
 * Cabinet; if not, write to the Free Software Foundation, Inc., 59 Temple Place, Suite 330,
 * Boston, MA 02111-1307 USA.
 *************************************************************************************************/


#include <tcutil.h>
#include <tcadb.h>
#include "myconf.h"


/* global variables */
const char *g_progname;                  // program name


/* function prototypes */
int main(int argc, char **argv);
static void usage(void);
static void printerr(TCADB *adb);
static int printdata(const char *ptr, int size, bool px);
static char *hextoobj(const char *str, int *sp);
static int runcreate(int argc, char **argv);
static int runinform(int argc, char **argv);
static int runput(int argc, char **argv);
static int runout(int argc, char **argv);
static int runget(int argc, char **argv);
static int runlist(int argc, char **argv);
static int runversion(int argc, char **argv);
static int proccreate(const char *name);
static int procinform(const char *name);
static int procput(const char *name, const char *kbuf, int ksiz, const char *vbuf, int vsiz,
                   int dmode);
static int procout(const char *name, const char *kbuf, int ksiz);
static int procget(const char *name, const char *kbuf, int ksiz, bool px, bool pz);
static int proclist(const char *name, int max, bool pv, bool px, const char *fmstr);
static int procversion(void);


/* main routine */
int main(int argc, char **argv){
  g_progname = argv[0];
  if(argc < 2) usage();
  int rv = 0;
  if(!strcmp(argv[1], "create")){
    rv = runcreate(argc, argv);
  } else if(!strcmp(argv[1], "inform")){
    rv = runinform(argc, argv);
  } else if(!strcmp(argv[1], "put")){
    rv = runput(argc, argv);
  } else if(!strcmp(argv[1], "out")){
    rv = runout(argc, argv);
  } else if(!strcmp(argv[1], "get")){
    rv = runget(argc, argv);
  } else if(!strcmp(argv[1], "list")){
    rv = runlist(argc, argv);
  } else if(!strcmp(argv[1], "version") || !strcmp(argv[1], "--version")){
    rv = runversion(argc, argv);
  } else {
    usage();
  }
  return rv;
}


/* print the usage and exit */
static void usage(void){
  fprintf(stderr, "%s: the command line utility of the abstract database API\n", g_progname);
  fprintf(stderr, "\n");
  fprintf(stderr, "usage:\n");
  fprintf(stderr, "  %s create name\n", g_progname);
  fprintf(stderr, "  %s inform name\n", g_progname);
  fprintf(stderr, "  %s put [-sx] [-dk|-dc] name key value\n", g_progname);
  fprintf(stderr, "  %s out [-sx] name key\n", g_progname);
  fprintf(stderr, "  %s get [-sx] [-px] [-pz] name key\n", g_progname);
  fprintf(stderr, "  %s list [-m num] [-pv] [-px] [-fm str] name\n", g_progname);
  fprintf(stderr, "  %s version\n", g_progname);
  fprintf(stderr, "\n");
  exit(1);
}


/* print error information */
static void printerr(TCADB *adb){
  fprintf(stderr, "%s: error\n", g_progname);
}


/* print record data */
static int printdata(const char *ptr, int size, bool px){
  int len = 0;
  while(size-- > 0){
    if(px){
      if(len > 0) putchar(' ');
      len += printf("%02X", *(unsigned char *)ptr);
    } else {
      putchar(*ptr);
      len++;
    }
    ptr++;
  }
  return len;
}


/* create a binary object from a hexadecimal string */
static char *hextoobj(const char *str, int *sp){
  int len = strlen(str);
  char *buf = tcmalloc(len + 1);
  int j = 0;
  for(int i = 0; i < len; i += 2){
    while(strchr(" \n\r\t\f\v", str[i])){
      i++;
    }
    char mbuf[3];
    if((mbuf[0] = str[i]) == '\0') break;
    if((mbuf[1] = str[i+1]) == '\0') break;
    mbuf[2] = '\0';
    buf[j++] = (char)strtol(mbuf, NULL, 16);
  }
  buf[j] = '\0';
  *sp = j;
  return buf;
}


/* parse arguments of create command */
static int runcreate(int argc, char **argv){
  char *name = NULL;
  for(int i = 2; i < argc; i++){
    if(!name && argv[i][0] == '-'){
      usage();
    } else if(!name){
      name = argv[i];
    } else {
      usage();
    }
  }
  if(!name) usage();
  int rv = proccreate(name);
  return rv;
}


/* parse arguments of inform command */
static int runinform(int argc, char **argv){
  char *name = NULL;
  for(int i = 2; i < argc; i++){
    if(!name && argv[i][0] == '-'){
      usage();
    } else if(!name){
      name = argv[i];
    } else {
      usage();
    }
  }
  if(!name) usage();
  name = tcsprintf("%s#mode=r", name);
  int rv = procinform(name);
  tcfree(name);
  return rv;
}


/* parse arguments of put command */
static int runput(int argc, char **argv){
  char *name = NULL;
  char *key = NULL;
  char *value = NULL;
  int dmode = 0;
  bool sx = false;
  for(int i = 2; i < argc; i++){
    if(!name && argv[i][0] == '-'){
      if(!strcmp(argv[i], "-dk")){
        dmode = -1;
      } else if(!strcmp(argv[i], "-dc")){
        dmode = 1;
      } else if(!strcmp(argv[i], "-sx")){
        sx = true;
      } else {
        usage();
      }
    } else if(!name){
      name = argv[i];
    } else if(!key){
      key = argv[i];
    } else if(!value){
      value = argv[i];
    } else {
      usage();
    }
  }
  if(!name || !key || !value) usage();
  int ksiz, vsiz;
  char *kbuf, *vbuf;
  if(sx){
    kbuf = hextoobj(key, &ksiz);
    vbuf = hextoobj(value, &vsiz);
  } else {
    ksiz = strlen(key);
    kbuf = tcmemdup(key, ksiz);
    vsiz = strlen(value);
    vbuf = tcmemdup(value, vsiz);
  }
  int rv = procput(name, kbuf, ksiz, vbuf, vsiz, dmode);
  tcfree(vbuf);
  tcfree(kbuf);
  return rv;
}


/* parse arguments of out command */
static int runout(int argc, char **argv){
  char *name = NULL;
  char *key = NULL;
  bool sx = false;
  for(int i = 2; i < argc; i++){
    if(!name && argv[i][0] == '-'){
      if(!strcmp(argv[i], "-sx")){
        sx = true;
      } else {
        usage();
      }
    } else if(!name){
      name = argv[i];
    } else if(!key){
      key = argv[i];
    } else {
      usage();
    }
  }
  if(!name || !key) usage();
  int ksiz;
  char *kbuf;
  if(sx){
    kbuf = hextoobj(key, &ksiz);
  } else {
    ksiz = strlen(key);
    kbuf = tcmemdup(key, ksiz);
  }
  int rv = procout(name, kbuf, ksiz);
  tcfree(kbuf);
  return rv;
}


/* parse arguments of get command */
static int runget(int argc, char **argv){
  char *name = NULL;
  char *key = NULL;
  bool sx = false;
  bool px = false;
  bool pz = false;
  for(int i = 2; i < argc; i++){
    if(!name && argv[i][0] == '-'){
      if(!strcmp(argv[i], "-sx")){
        sx = true;
      } else if(!strcmp(argv[i], "-px")){
        px = true;
      } else if(!strcmp(argv[i], "-pz")){
        pz = true;
      } else {
        usage();
      }
    } else if(!name){
      name = argv[i];
    } else if(!key){
      key = argv[i];
    } else {
      usage();
    }
  }
  if(!name || !key) usage();
  int ksiz;
  char *kbuf;
  if(sx){
    kbuf = hextoobj(key, &ksiz);
  } else {
    ksiz = strlen(key);
    kbuf = tcmemdup(key, ksiz);
  }
  name = tcsprintf("%s#mode=r", name);
  int rv = procget(name, kbuf, ksiz, px, pz);
  tcfree(name);
  tcfree(kbuf);
  return rv;
}


/* parse arguments of list command */
static int runlist(int argc, char **argv){
  char *name = NULL;
  int max = -1;
  bool pv = false;
  bool px = false;
  char *fmstr = NULL;
  for(int i = 2; i < argc; i++){
    if(!name && argv[i][0] == '-'){
      if(!strcmp(argv[i], "-m")){
        if(++i >= argc) usage();
        max = atoi(argv[i]);
      } else if(!strcmp(argv[i], "-pv")){
        pv = true;
      } else if(!strcmp(argv[i], "-px")){
        px = true;
      } else if(!strcmp(argv[i], "-fm")){
        if(++i >= argc) usage();
        fmstr = argv[i];
      } else {
        usage();
      }
    } else if(!name){
      name = argv[i];
    } else {
      usage();
    }
  }
  if(!name) usage();
  name = tcsprintf("%s#mode=r", name);
  int rv = proclist(name, max, pv, px, fmstr);
  tcfree(name);
  return rv;
}


/* parse arguments of version command */
static int runversion(int argc, char **argv){
  int rv = procversion();
  return rv;
}


/* perform create command */
static int proccreate(const char *name){
  TCADB *adb = tcadbnew();
  if(!tcadbopen(adb, name)){
    printerr(adb);
    tcadbdel(adb);
    return 1;
  }
  bool err = false;
  if(!tcadbclose(adb)){
    printerr(adb);
    err = true;
  }
  tcadbdel(adb);
  return err ? 1 : 0;
}


/* perform inform command */
static int procinform(const char *name){
  TCADB *adb = tcadbnew();
  if(!tcadbopen(adb, name)){
    printerr(adb);
    tcadbdel(adb);
    return 1;
  }
  bool err = false;
  printf("record number: %llu\n", (unsigned long long)tcadbrnum(adb));
  printf("size: %llu\n", (unsigned long long)tcadbsize(adb));
  if(!tcadbclose(adb)){
    printerr(adb);
    err = true;
  }
  tcadbdel(adb);
  return err ? 1 : 0;
}


/* perform put command */
static int procput(const char *name, const char *kbuf, int ksiz, const char *vbuf, int vsiz,
                   int dmode){
  TCADB *adb = tcadbnew();
  if(!tcadbopen(adb, name)){
    printerr(adb);
    tcadbdel(adb);
    return 1;
  }
  bool err = false;
  switch(dmode){
  case -1:
    if(!tcadbputkeep(adb, kbuf, ksiz, vbuf, vsiz)){
      printerr(adb);
      err = true;
    }
    break;
  case 1:
    if(!tcadbputcat(adb, kbuf, ksiz, vbuf, vsiz)){
      printerr(adb);
      err = true;
    }
    break;
  default:
    if(!tcadbput(adb, kbuf, ksiz, vbuf, vsiz)){
      printerr(adb);
      err = true;
    }
    break;
  }
  if(!tcadbclose(adb)){
    if(!err) printerr(adb);
    err = true;
  }
  tcadbdel(adb);
  return err ? 1 : 0;
}


/* perform out command */
static int procout(const char *name, const char *kbuf, int ksiz){
  TCADB *adb = tcadbnew();
  if(!tcadbopen(adb, name)){
    printerr(adb);
    tcadbdel(adb);
    return 1;
  }
  bool err = false;
  if(!tcadbout(adb, kbuf, ksiz)){
    printerr(adb);
    err = true;
  }
  if(!tcadbclose(adb)){
    if(!err) printerr(adb);
    err = true;
  }
  tcadbdel(adb);
  return err ? 1 : 0;
}


/* perform get command */
static int procget(const char *name, const char *kbuf, int ksiz, bool px, bool pz){
  TCADB *adb = tcadbnew();
  if(!tcadbopen(adb, name)){
    printerr(adb);
    tcadbdel(adb);
    return 1;
  }
  bool err = false;
  int vsiz;
  char *vbuf = tcadbget(adb, kbuf, ksiz, &vsiz);
  if(vbuf){
    printdata(vbuf, vsiz, px);
    if(!pz) putchar('\n');
    tcfree(vbuf);
  } else {
    printerr(adb);
    err = true;
  }
  if(!tcadbclose(adb)){
    if(!err) printerr(adb);
    err = true;
  }
  tcadbdel(adb);
  return err ? 1 : 0;
}


/* perform list command */
static int proclist(const char *name, int max, bool pv, bool px, const char *fmstr){
  TCADB *adb = tcadbnew();
  if(!tcadbopen(adb, name)){
    printerr(adb);
    tcadbdel(adb);
    return 1;
  }
  bool err = false;
  if(fmstr){
    TCLIST *keys = tcadbfwmkeys2(adb, fmstr, max);
    for(int i = 0; i < tclistnum(keys); i++){
      int ksiz;
      const char *kbuf = tclistval(keys, i, &ksiz);
      printdata(kbuf, ksiz, px);
      if(pv){
        int vsiz;
        char *vbuf = tcadbget(adb, kbuf, ksiz, &vsiz);
        if(vbuf){
          putchar('\t');
          printdata(vbuf, vsiz, px);
          tcfree(vbuf);
        }
      }
      putchar('\n');
    }
    tclistdel(keys);
  } else {
    if(!tcadbiterinit(adb)){
      printerr(adb);
      err = true;
    }
    int ksiz;
    char *kbuf;
    int cnt = 0;
    while((kbuf = tcadbiternext(adb, &ksiz)) != NULL){
      printdata(kbuf, ksiz, px);
      if(pv){
        int vsiz;
        char *vbuf = tcadbget(adb, kbuf, ksiz, &vsiz);
        if(vbuf){
          putchar('\t');
          printdata(vbuf, vsiz, px);
          tcfree(vbuf);
        }
      }
      putchar('\n');
      tcfree(kbuf);
      if(max >= 0 && ++cnt >= max) break;
    }
  }
  if(!tcadbclose(adb)){
    if(!err) printerr(adb);
    err = true;
  }
  tcadbdel(adb);
  return err ? 1 : 0;
}


/* perform version command */
static int procversion(void){
  printf("Tokyo Cabinet version %s (%d:%s)\n", tcversion, _TC_LIBVER, _TC_FORMATVER);
  printf("Copyright (C) 2006-2008 Mikio Hirabayashi\n");
  return 0;
}



// END OF FILE
